### ISSUE TITLE
Unify text-mode and math-mode LaTeX parsing

### ISSUE THREAD

**Issue by Happypig375** [verybadcat/CSharpMath#58]
**Is your feature request related to a problem? Please describe.**
Currently, the APIs for CSharpMath are not intuitive at all.
For example, to convert MathList->TeX, one has to call `MathListBuilder.MathListToString` which is a static method (not easy to discover), and its name doesn't imply anything about the result being a TeX string at all.

**Describe the solution you'd like**
1. Streamline the use of CSharpMath APIs.
```cs
SkiaSharp.SKCanvas canvas = …;
\\frac{1}{2}
.TeXToMathAtoms()
.ToDisplay()
.ToOutput(canvas);
```

2. Replace mutable global variables with settings objects.
e.g. `CSharpMath.Atoms.MathAtoms` contain multiple dictionaries that act as hidden dependencies. They should be made obvious.

3. Reduce legacy boilerplate code from iosMath.
e.g. Why do some math atoms have their respective interfaces? It's not like MathAtoms will have alternate implementations, they are just data containers. If they are to be extended, they can be inherited.
e.g. Reduce boilerplate folders. They scare new contributors. Most folders in `CSharpMath/` contain few files anyway.
e.g. Remove CSharpMath.Enumerations.MathAtomType. Use pattern matching on types instead.

4. Simplify the NuGet chain. It's not obvious what `CSharpMath.Rendering` does based on its name. `CSharpMath`, `CSharpMath.Editor` and `CSharpMath.Rendering` should be merged into one. Typography should be integrated directly into CSharpMath because

.NET font reader library name | Can read math layout info for glyphs?
-|-
[NRasterizer](https://github.com/vidstige/NRasterizer) | :x:
[SixLabors.Fonts](https://github.com/SixLabors/Fonts) | :x:
[SharpFont](https://github.com/MikePopoloski/SharpFont) | :x:
[Typography](https://github.com/LayoutFarm/Typography) | :heavy_check_mark:

###### Source: https://github.com/Jolg42/awesome-typography#c-2

It's not like there is or will be any other implementations for font reading of math fonts.

**Describe alternatives you've considered**
Continuing the current API and result in a degraded user experience.

**Difficulty: How difficult would it be? (Trivial, Very Easy, Easy, Moderate, Hard, Very Hard, Tedious, Backbreaking)**
Very hard. A lot of time will be needed to achieve this.

**Additional context**
#57

## Maintainer triage (2026-08-30): remaining scope

The original umbrella proposal mixed several independent decisions. The public painter API and documentation now cover the discoverability goal; legacy boilerplate was reduced; and merging all frontend NuGet packages was rejected. The remaining actionable architectural item is to unify text-mode and math-mode LaTeX parsing so that `\text` can switch modes without maintaining two command dictionaries and two syntax trees.

### Acceptance criteria

- A single parsing architecture can enter text mode from math mode and return to math mode.
- Commands shared by text and math have one definition and consistent serialization.
- Existing math and text round-trip tests continue to pass, with new mixed-mode tests.
- Frontend packages and the painter-based public API remain independently consumable.

**Comment by Smurf-IV** [comment]
Please can you keep the "Core" of the CSharpMath, out of the need for "Rendering and editor" as this allows it to be used for other frameworks, i.e. Services only.

**Comment by charlesroddie** [comment]
What are these services and how would CSharpMath be used without rendering?

**Comment by Smurf-IV** [comment]
@charlesroddie 
- .Net Core 2
- Used for Verification of the Latex Format (Hence my questions/problems about round tripping)
- Then transferred Via the` MathList etc.` - using internal methods - into different usage and logic trees, dependent on wrapped metadata and context.

**Comment by Happypig375** [comment]
Other frameworks can still be used.
![image](https://user-images.githubusercontent.com/19922066/64924363-6791cf80-d816-11e9-9e5e-5dc5c45076f4.png)

**Comment by Happypig375** [comment]
Editing capabilities will be baked into MathAtoms.

**Comment by charlesroddie** [comment]
> Streamline the use of CSharpMath APIs.

Very good but not sure about the example. I don't think that we should add extension methods to `string`! If you are creating object B from A then `B.FromA(a:A)` is often better because usually A can be defined without reference to B but not vice versa.

> Replace mutable global variables with settings objects.

Sounds good.

> Reduce legacy boilerplate code from iosMath.

Great.

> Simplify the NuGet chain. It's not obvious what CSharpMath.Rendering does based on its name. CSharpMath, CSharpMath.Editor and CSharpMath.Rendering should be merged into one. Typography should be integrated directly into CSharpMath because

That's not a significant problem for users. If the projects have a natural structure where later projects naturally depend on earlier ones, then having things in separate projects enforces that ordering, unlike C# code in the same project. I don't know if this is the case with the current CSharpMath projects though.

**Comment by Happypig375** [comment]
An update since the original post:
1. Streamline the use of CSharpMath APIs: Better solved by documentation as in https://github.com/verybadcat/CSharpMath#major-processes-of-drawing-latex. Extension methods would result in pollution as noted by @charlesroddie.
2. Replace mutable global variables with settings objects: To do. This is required to support `\def`, `\DeclareMathOperator` etc.
3. Reduce legacy boilerplate code from iosMath: Already done as part of the 0.4 update.
4. Simplify the NuGet chain: Not a good idea as noted by @charlesroddie.

New items:
5. Merge the entirety of `CSharpMath.Rendering.Text` into `CSharpMath.Atom`. In LaTeX, `\text` enables text-mode inside math-mode. With a separate implementation of `CSharpMath.Rendering.Text`, this is impossible. It is also undesirable to maintain two LaTeX parsers and two LaTeX trees at the same time. Moreover, this results in difference functionality between `CSharpMath.Atom` and `CSharpMath.Rendering.Text`, e.g. `CSharpMath.Atom` supports `\colorbox` while `CSharpMath.Rendering.Text` supports `\fontsize`, when such functionality should be present on both implementations. With #143, the same parser can be reused for both math-mode and text-mode with swappable command dictionaries.

An open question is whether to support (MathList->Text LaTeX) on top of (MathList->Math LaTeX). Math LaTeX seems enough for everyday use.

The remaining items will be done as "Parser refactor (part 2)", blocked on #143.

**Comment by Happypig375** [comment]
Note for implementing display math inside text: as seen in http://www.dfcd.net/articles/latex/latex.html, we can add a `center` environment that helps the implementation.
